### ISSUE TITLE
Add recurring item by day/week capability

### ISSUE THREAD

**Issue by jmwerner** [derek-damron/budgetr#26]
e.g. "Gettin paaaaid every 14 days". Pardon me if I missed it somewhere.

**Comment by derek-damron** [comment]
You didn't miss it nowhere. :smile_cat:

## Quick and dirty solution

Create a bunch of items manually using a start date.  For example, creating items for the next 4 pay days:

**Input**

```
library(budgetr)

# Get a vector of dates for the next 4 pay days
paaaaid_dates <- as.Date("2016-09-16") + seq.int(0, by = 14, length.out = 4)

# Create the pay days
for (i in 1:length(paaaaid_dates)) {
    # Create name of the item object
    loop_assign <- paste0("paaaaid_dates_", format(paaaaid_dates[i], format="%Y%m%d"))
    # Create the item
    assign( loop_assign
          , create_item( name = loop_assign
                       , amount = 666
                       , day = paaaaid_dates[i]
                       , recurring = FALSE
                       ))
    # Drop the loop objects
    rm(i, loop_assign)
}

# Check those items
get_items(pattern = "paaaaid_dates")
```

**Output**

```{r}
$paaaaid_dates_20160916
item (budgetr)
                   name amount        day recurring
 paaaaid_dates_20160916    666 2016-09-16     FALSE

$paaaaid_dates_20160930
item (budgetr)
                   name amount        day recurring
 paaaaid_dates_20160930    666 2016-09-30     FALSE

$paaaaid_dates_20161014
item (budgetr)
                   name amount        day recurring
 paaaaid_dates_20161014    666 2016-10-14     FALSE

$paaaaid_dates_20161028
item (budgetr)
                   name amount        day recurring
 paaaaid_dates_20161028    666 2016-10-28     FALSE
```

## Taking the long view

As relevant as this is to me I've putting this off because I haven't thought of an efficient way to implement it yet in the current setup.  It seems like in a case like this where you want something that is recurring outside of the monthly time frame you'd need:

- The time frame for recurrence (e.g. "2 weeks")
- The first "effective" date (e.g. "2016-09-16" or "This Friday and not that Friday")

One way to implement this in the current setup would be to pass a list argument to `recurring` like so:

```{r}
create_item( ...
           , recurring = list(time = "2 weeks", start = as.Date("2016-09-16"))
           )
```

I think this would work but it seems a bit unwieldy.  It might be better to use this as an excuse to revamp the `recurring` argument into defaulting to "monthly" as opposed to the current `TRUE`/`FALSE` setup.  I have a suspicion that this would require some rework of the `create_*` and `update_*` internals but this does seem like one of the bigger "next steps" for making *budgetr* more flexible.

Whacha thank, @jmwerner?

**Comment by derek-damron** [comment]
After #31 we should be able to implement this.

Rather than the formulation above, you'll be able to create a 2-week recurring items by specifying the exact start date and a non-logical recurring argument.  For example:

```{r}
create_item( name = "Gettin paaaaid every 14 days"
           , amount = 666
           , day = "2016-09-16"
           , recurring = "2 weeks"
           )
```

That sound like whacha need, @jmwerner?

**Comment by jmwerner** [comment]
Most def, that would be fantastic! I’d even recommend having the “recurring” argument take the number of days for one less text field to parse, but having it optionally take weeks/fortnights/etc is probably handled well by lubridate regardless

**Comment by derek-damron** [comment]
I'm reluctant to have `recurring` be the number of days because that makes monthly expenses due on the same day of each month (e.g. rent) incredibly difficult for the user to implement without adding some type of "interval" argument.

**Comment by jmwerner** [comment]
Right, those represent two totally different recurrence types. That being said, maybe that behavior could be built in, circa:

One thang (assuming recurring is false by default)

```
create_item( name = "New Purchase"
                    , amount = 666
                    , date = "2016-09-16"
                    )
```

Monthly thang (repeats every month on whatever date - or day with the current setup - it's given)

```
create_item( name = "Rent"
                    , amount = 666
                    , date = "2016-09-16"
                    , recurring = "monthly",
                    )
```

Biweekly thang ("14 days" or "2 weeks" or whatever passed to recurring)

```
create_item( name = "Mas Dinero"
                    , amount = 666
                    , date = "2016-09-16"
                    , recurring = "14 days",
                    )
```

I'm partial to even forcing the user to specify that other recurring events are to recur monthly to reinforce that delineation. Just some food for thought.

**Comment by derek-damron** [comment]
I think we're on the same page!

Here are the different `recurring` arguments I had in mind for now.  Can you think of any other strings that might be useful?

- `FALSE`: one-time item
- Days
  - `"daily"` or `"1 day"`: repeat every day
  - `"X days"`: repeat every X days
- Weeks
  - `"weekly"` or `"1 week"`: repeat every week
  - `"X weeks"`: repeat every X weeks
- Months
  - `"monthly"` or `"1 month"`: repeat every month
  -  `"X months"`: repeat every X months
- Years
  - `"yearly"` or `"1 year"`: repeat every year
  - `"X years"`: repeat every X years

**Comment by jmwerner** [comment]
![i like it](http://i1248.photobucket.com/albums/hh488/ifoundcallie/101379-Jim-Carrey-I-like-it-alot-gif-GflM.gif)

Can't think of any others, that should do it!

**Comment by derek-damron** [comment]
This is basically done in the dev-v0.2.2 branch but I want to make a small tweak to how an item's `day` interacts with a budget's `start` and `end` (#33) before wrapping up the milestone and merging to master.  Check it out!

```r
paycheck <- create_item( name = "Paycheck"
                       , amount = 1000
                       , day = "2016-01-01"
                       , recurring = "monthly"
                       )
rent <- create_item( name = "Rent"
                   , amount = -500
                   , day = "2016-01-05"
                   , recurring = "monthly"
                   )
groceries <- create_item( name = "Groceries"
                        , amount = -100
                        , day ="2015-12-15"
                        , recurring = "2 weeks"
                        )
xmas_gifts <- create_item( name = "Christmas Gifts"
                         , amount = -500
                         , day = "2015-12-20"
                         )
my_schedule <- create_schedule(paycheck, rent, groceries, xmas_gifts)
my_budget <- create_budget( my_schedule
                          , start = "2015-12-15"
                          , end = "2016-03-02"
                          , initial = 1000
                          )
my_budget
# budget (budgetr)
#        date            name amount balance
#  2015-12-15  Initial Amount   1000    1000
#  2015-12-15       Groceries   -100     900
#  2015-12-20 Christmas Gifts   -500     400
#  2015-12-29       Groceries   -100     300
#  2016-01-01        Paycheck   1000    1300
#  2016-01-05            Rent   -500     800
#  2016-01-12       Groceries   -100     700
#  2016-01-26       Groceries   -100     600
#  2016-02-01        Paycheck   1000    1600
#  2016-02-05            Rent   -500    1100
#  2016-02-09       Groceries   -100    1000
#  2016-02-23       Groceries   -100     900
#  2016-03-01        Paycheck   1000    1900

**Comment by jmwerner** [comment]
![idig](http://www.brainlesstales.com/images/2010/Jul/dig-it.jpg)

**Comment by derek-damron** [comment]
It has been merged into master.  I'll let you close out this issue once you check that it's working on your end. :smile_cat:

**Comment by jmwerner** [comment]
![me gusta](http://67.media.tumblr.com/tumblr_lj0f78lDt91qibz0jo1_r1_500.png)
